### ISSUE TITLE
Performance issue with 64bit on CPU

### ISSUE THREAD

**Issue by inversecrime** [google/jax#20616]
### Description

Hi,
I encountered a performance issue when enabling 64bit computations, and I would be glad if someone could help me here - is this a known issue, how do i fix it, etc.
Below the example.

```
import time
from contextlib import contextmanager
from functools import partial
from typing import Iterator

import jax
import jax.numpy as jnp

jax.config.update("jax_enable_x64", True)  # every option takes approximately the same time, as expected, when this is left out


@contextmanager
def measure_time(name: str) -> Iterator[None]:
    start = time.time()
    yield
    stop = time.time()
    print(f"Task [{name}] took {stop - start:.2f} s.")


def g(x, a, b, c, inner_func):
    d = jax.vmap(jax.vmap(
        jax.vmap(
            jax.vmap(
                inner_func,
                in_axes=(None, None, 0),
                out_axes=0
            ),
            in_axes=(None, 0, None),
            out_axes=0
        )
    ))(x, a, b)
    return jnp.einsum("ij,ij...->i...", c, d)


f1 = lambda x: 1 + jnp.sum(jnp.sin(jnp.pi * x)**2)  # f1 evaluates to 1.0 for x = 0.0, same as f2
f2 = lambda x: jnp.asarray(1.0)

inner_func1 = lambda x, u, v: f1(x) * u @ v
inner_func2 = lambda x, u, v: f1(x) * (u @ v)
inner_func3 = lambda x, u, v: f2(x) * u @ v
inner_func4 = lambda x, u, v: f2(x) * (u @ v)

g1 = partial(g, inner_func=inner_func1)
g2 = partial(g, inner_func=inner_func2)
g3 = partial(g, inner_func=inner_func3)
g4 = partial(g, inner_func=inner_func4)


x = jnp.zeros((13720, 57, 3))
a = jnp.zeros((13720, 57, 35, 3))
b = jnp.zeros((13720, 57, 35, 3))
c = jnp.zeros((13720, 57))

with measure_time("g1"):
    jax.block_until_ready(jax.jit(g1)(x, a, b, c))
with measure_time("g2"):
    jax.block_until_ready(jax.jit(g2)(x, a, b, c))  # much much slower than all other options when 64bit is enabled
with measure_time("g3"):
    jax.block_until_ready(jax.jit(g3)(x, a, b, c))
with measure_time("g4"):
    jax.block_until_ready(jax.jit(g4)(x, a, b, c))
```

All 4 functions g1-g4 produce the same result (for `x = 0.0`), but g2 takes significantly longer than all other options.

Of course, in real use cases, the inputs are not zero.

g1 / g2
and
g3 / g4
are mathematically equivalent.

I tried to keep this code chunk as small as possible without making it hard to understand what's going on.
This code is used for finite element computation, so this seemingly random function g has actual meaning to it.

I would appreciate any help!

### System info (python version, jaxlib version, accelerator, etc.)

python 3.12
jax 0.4.26
jaxlib 0.4.26
CPU (AMD)

**Comment by inversecrime** [comment]
Is it possible that this is a RAM issue? It almost seems like it - with smaller arrays, even with x64 everything looks ok.

But then again, I do not see (even with looking at the jaxpr) how the second option should be more expensive in terms of RAM than the others.

**Comment by inversecrime** [comment]
I think I tracked down the issue, and it looks like this even happens with 32bit numbers.
Consider the following example, where the second function almost uses twice as much memory as the first one.
The only thing different is the order of multiplication and dot product - as far as I know, this should not affect memory at all.
Is there any way to check this?

````
import time
from contextlib import contextmanager
from typing import Iterator

import jax
import jax.numpy as jnp


@contextmanager
def measure_time(name: str) -> Iterator[None]:
    start = time.time()
    yield
    stop = time.time()
    print(f"Task [{name}] took {stop - start:.2f} s.")


key = jax.random.PRNGKey(0)
(subkey1, subkey2, subkey3, subkey4) = jax.random.split(key, 4)
x = jax.random.uniform(subkey1, (15000, 50, 1, 1))
a = jax.random.uniform(subkey2, (15000, 50, 35, 5))
b = jax.random.uniform(subkey3, (15000, 50, 35, 5))
c = jax.random.uniform(subkey4, (15000, 50))


def h1(x, a, b, c):
    a = x * a
    d = jnp.einsum("...ji,...ki->...jk", a, b)
    return jnp.einsum("ij,ij...->i...", c, d)


def h2(x, a, b, c):
    d = jnp.einsum("...ji,...ki->...jk", a, b)
    d = x * d
    return jnp.einsum("ij,ij...->i...", c, d)


with measure_time("h1"):
    y = jax.jit(h1)(x, a, b, c)
    jax.block_until_ready(y)

with measure_time("h2"):
    y = jax.jit(h2)(x, a, b, c)
    jax.block_until_ready(y)
````

**Comment by jakevdp** [comment]
Thanks for the report. It looks like the compiler is able to recognize the constant expression in one case, but not in the other.

I don't think I'd consider this a bug: we can't expect the compiler to be perfect, and a long-enough sequence of operations may evade its rewrite rules. But we might raise this issue in http://github.com/openxla/xla: perhaps there's a way to update the compiler's heuristics to catch this case.

**Comment by inversecrime** [comment]
Thanks for the answer. I don't have any experience with XLA per se, so I wouldn't know where to start there.
Also, I don't really know how to observe RAM usage of a compiled function reliably.

Currently I think the problem is that in this function

```
def f(x, a, b, c):
    d = jnp.einsum("...ji,...ki->...jk", a, b)
    d = x * d
    return d
```

the multiplication `x * d` is not performed in-place even though it would technically be possible.
Do you know how it would be possible to confirm this?

**Comment by inversecrime** [comment]
I mean, it's just a suspicion.
But why else would it need so much more memory than
```
def f(x, a, b, c):
    a = x * a
    d = jnp.einsum("...ji,...ki->...jk", a, b)
    return d
```
But then again, I don't even know how to veryify that this memory problem is "real" and not just a "on my machine" problem.

**Comment by jakevdp** [comment]
We can see what the compiler is doing with these functions using [ahead of time lowering](https://jax.readthedocs.io/en/latest/aot.html). For example:
```
h1_lowered = jax.jit(h1).lower(x, a, b, c).compile()
h2_lowered = jax.jit(h2).lower(x, a, b, c).compile()

# print compiled HLO
print(h1_lowered.as_text())
print(h2_lowered.as_text())

# print compiled HLO
print(h1_lowered.cost_analysis())
print(h2_lowered.cost_analysis())
```
If you inspect the outputs, you see that the second function has a much larger "bytes accessed" in the cost analysis. I believe this is due to the compiler not having any heuristic to fuse the operations when they are performed in this order, but I'm not certain of that interpretation.

**Comment by jakevdp** [comment]
I think I understand the difference: the expensive allocation is the output of the `einsum`. In `h1`, the input to the einsum is an internal buffer (the output of `x * a`) and therefore can be overwritten. In `h2`, both inputs are user-level buffers, and so neither are overwritable by default and the compiler must allocate a new large buffer for the output.

You could probably address this by using `donate_argnums` to tell the compiler that one or more of the input buffers is overwriteable (but of course you would not be able to use those donated buffers in subsequent computations).

If this operation is done as part of a larger program, then it's likely that the inputs will already be temporary buffers, and this microbenchmark will not be reflective of the overall performance of the larger program.

Hope that's clear!

**Comment by jakevdp** [comment]
Closing, as I think this is working as expected.

**Comment by inversecrime** [comment]
Thanks for helping @jakevdp. I agree, _IF_ this is an issue, it's definitely not JAX, but whatever the compiler does.

In the meantime I played around with this a bit and noticed two important things (Your comments have been very helpful):

1. In the code I posted in this comment https://github.com/google/jax/issues/20616#issuecomment-2041276158, there is no problem at all. Both functions need exactly the same amount of memory (I think, not entirely sure though). If you simply put a `del y` in between the function calls, everything works as expected. That's on me, sorry for that.

2. The original issue still persists, I'm having huge performance drops (which I still think has to do with memory usage) in my actual code, which is still reproducable with the original code I posted in this issue (in which the function `g2` uses almost twice as much memory as the other options).

If I find the actual reason for my problem (and it has something to do with jax), I'll probably just open a new issue, if that is ok.
